text/2021-12-29-openapi-to-typescript-api-client-and-forms.md

```diff
@@ -0,0 +1,458 @@
+# OpenAPI to TypeScript API Client and Forms
+
+## Summary
+
+Use the OpenAPI specification to generate a TypeScript API client and forms.
+
+## Motivation
+
+Ref: https://github.com/chaos-mesh/chaos-mesh/issues/2615.
+
+Currently, the chaos types are split between the front-end and back-end, and we
+can‘t reuse types that have already been defined.
+
+You can find these **extra** type definitions in https://github.com/chaos-mesh/chaos-mesh/blob/release-2.1/ui/app/src/components/NewExperimentNext/data/types.ts.
+
+This leads to several problems:
+
+1. The front-end needs to be manually synchronized when there is an update to the
+   type definition.
+2. As `1` continues to emerge, there are more and more things to maintain manually.
+
+Considering our existing maintenance cost and past experience, we are not able
+to synchronize the corresponding descriptions to the front-end in a timely manner.
+
+This also works against those who want to contribute to Chaos Mesh, and it would
+be great if only modifying the API will synchronize the changes to the UI.
+
+So the best solution for now is to automate the generation of the schemas needed
+for the front-end.
+
+## Detailed design
+
+I will split the details into two parts:
+
+- The first part is the generation of the TypeScript schemas.
+- The second part is how to use the generated files to produce a forms skeleton.
+
+### Generate TypeScript Schemas
+
+Luckily, there has serval tools can help us to generate the schemas, all of them
+has pros and cons, finally I choose the
+[OpenAPITools/openapi-generator](https://github.com/OpenAPITools/openapi-generator),
+because:
+
+- Pros
+  - It's official.
+  - It's a very popular tool and has a huge community.
+- Cons
+  - The nodejs package is just a wrapper of jar, so JRE is needed.
+
+Other tries:
+
+- [openapi-typescript](https://www.npmjs.com/package/openapi-typescript)
+- [openapi-typescript-codegen](https://www.npmjs.com/package/openapi-typescript-codegen)
+
+Both of these tools also have it's pros and cons, as opposite of above described.
+
+Then we can use create a new package to handle the generation:
+
+```sh
+mkdir -p ui/packages/openapi
+cd ui/package/openapi
+yarn init
+```
+
+Install dependencies and add the `generate` script:
+
+```json
+{
+  "scripts": {
+    "generate": "export TS_POST_PROCESS_FILE='../../node_modules/.bin/prettier --write'; openapi-generator-cli generate -c openapiconfig.json -i ../../../pkg/dashboard/swaggerdocs/swagger.yaml -g typescript-axios -o ../../app/src/openapi --enable-post-process-file"
+  }
+}
+```
```

Comment:
Currently, there is still a problem here that I haven't had time to explain is **$ref siblings aren't supported in OpenAPI v2**. For example:

```yaml
attr:
  $ref: '#/definitions/v1alpha1.AttrOverrideSpec'
  description: |-
    Attr defines the overrided attribution
    ui:form:action=attrOverride
    +optional
  type: object
```

The above definitions will convert to:

```ts
/**
 *
 * @type {V1alpha1AttrOverrideSpec}
 * @memberof V1alpha1IOChaosSpec
 */
attr?: V1alpha1AttrOverrideSpec
```

The `description` will be lost 😢. After searching, I found a solution in: https://stackoverflow.com/questions/33629750/swagger-schema-properties-ignored-when-using-ref-why, I need to translate the definitions into this form to solve the problem:

```yaml
attr:
  allOf:
    - $ref: '#/definitions/v1alpha1.AttrOverrideSpec'
  description: |-
    Attr defines the overrided attribution
    ui:form:action=attrOverride
    +optional
  type: object
```

To do this I need to make a copy of `swagger.yaml` and then do the inline process. So I can't move this command to the workspaces root because it would break the integrity of the package. I will supplement this part into the RFC.